code/scripting/api/libs/async.cpp

```diff
@@ -214,6 +216,60 @@ ADE_FUNC(run,
 		l_Promise.Set(runAsyncCoroutine(std::move(coroutine), executor.getExecutor(), std::move(context))));
 }
 
+ADE_FUNC(awaitRunOnFrame,
+	l_Async,
+	"function() => any body",
+	"Runs an asynchronous function in an OnFrameExecutor context and busy-waits for the coroutine to finish. "
+	"Inside this function you can use async.await to suspend the function until a promise resolves. "
+	"This is useful for cases where you need a scripting process to run over multiple frames, even when "
+	"the engine is not in a stable game state (such as showing animations during game state switches, etc.).",
+	"any... | nil",
+	"The result of the function body or nil if the function errored")
+{
+	luacpp::LuaFunction body;
+	std::shared_ptr<executor::IExecutionContext> context;
+
+	if (!ade_get_args(L, "u", &body)) {
+		return ADE_RETURN_NIL;
+	}
+
+	context = executor::GameStateExecutionContext::captureContext();
+
+
+	auto coroutine = luacpp::LuaThread::create(L, body);
+	coroutine.setErrorCallback([](lua_State*, lua_State* thread) {
+		LuaError(thread);
+		return true;
+	});
+
+	auto promise = runAsyncCoroutine(std::move(coroutine), executor::OnFrameExecutor, std::move(context));
+
+	int screen_id = gr_save_screen();
+
+	//Keep drawing pretend-frames until we have the promise resolved
+	while (promise.isValid() && !(promise.isResolved() || promise.isErrored())){
```

Comment:
This loop doesn't handle multi packets, which may or may not be an issue later on.  So... we'll see if it does later on.

Reply:
This was intentional, actually.
Mostly because we want to 100% avoid the game being able to switch states during this, as well as explicitly allowing for something like "wait _now_ to be done loading stuff before you're allowed to proceed", which also _must_ make multi wait.
It's possible to add a parameter to the function, but the full-blocking mode must be an option, at least.
Plus, a few popups do it this way, so it can't be that much of a dealbreaker

Reply:
Well, I raise this point because it is probable the function will be (ab)used in a different way in the future by modders, so I want to see if there's anything we can do to prevent a crash or freeze.